overlord/devicestate/handlers_gadget.go

```diff
@@ -210,11 +212,96 @@ func (m *DeviceManager) doUpdateGadgetAssets(t *state.Task, _ *tomb.Tomb) error
 	return snapstate.FinishTaskWithRestart(t, state.DoneStatus, restart.RestartSystem, nil)
 }
 
-func (m *DeviceManager) updateGadgetCommandLine(t *state.Task, st *state.State, isUndo bool) (updated bool, err error) {
-	snapsup, err := snapstate.TaskSnapSetup(t)
+// fromSystemOption tells us if t was created when setting a system
+// option for the kernel command line.
+func fromSystemOption(t *state.Task) (bool, error) {
+	for _, param := range []string{"cmdline-append", "dangerous-cmdline-append"} {
+		var value string
+		err := t.Get(param, &value)
+		if err == nil {
+			return true, nil
+		}
+		if !errors.Is(err, state.ErrNoState) {
+			return false, err
+		}
+	}
+
+	return false, nil
+}
+
+// kernelCommandLineAppendArgs returns extra arguments that we want to
+// append to the kernel command line, searching first by looking at
+// the task, and if not found, looking at the current configuration
+// options. One thing or the other could happen depending on whether
+// this is a task created when setting a kernel option or by gadget
+// installation.
+func kernelCommandLineAppendArgs(tsk *state.Task, tr *config.Transaction,
+	taskParam string) (string, error) {
+
+	var value string
+	err := tsk.Get(taskParam, &value)
+	if err == nil {
+		return value, nil
+	}
+	if !errors.Is(err, state.ErrNoState) {
+		return "", err
+	}
+
+	var option string
+	switch taskParam {
+	case "cmdline-append":
+		option = "system.kernel.cmdline-append"
+	case "dangerous-cmdline-append":
+		option = "system.kernel.dangerous-cmdline-append"
+	default:
+		return "", fmt.Errorf("internal error, unexpected task parameter %q", taskParam)
+	}
+	if err := tr.Get("core", option, &value); err != nil && !config.IsNoOption(err) {
+		return "", err
+	}
+
+	return value, nil
+}
+
+func buildAppendedKernelCommandLine(t *state.Task, gd *gadget.GadgetData) (string, error) {
+	st := t.State()
+	tr := config.NewTransaction(st)
+	rawCmdlineAppend, err := kernelCommandLineAppendArgs(t, tr, "cmdline-append")
 	if err != nil {
-		return false, err
+		return "", err
+	}
+	// Validation against allow list has already happened in
+	// configcore, but the gadget might have changed, so we check
+	// again and filter any unallowed argument.
+	cmdlineAppend, forbidden := gadget.FilterKernelCmdline(rawCmdlineAppend, gd.Info.KernelCmdline.Allow)
+	if forbidden != "" {
+		warnMsg := fmt.Sprintf("%q is not allowed by the gadget and has been filtered out from the kernel command line", forbidden)
+		logger.Noticef(warnMsg)
+		t.Logf(warnMsg)
+	}
+
+	// Dangerous extra cmdline only considered for dangerous models
+	deviceCtx, err := DeviceCtx(st, nil, nil)
```

Comment:
for consistency this should pass in t, see preexisting usages of DeviceCtx in this file or the function should take a DeviceContext as one is available in the caller?

Reply:
Changed (I assume you mean `st`)